miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+
+        ex = Example(vm, addr)
+        ex.mystruct.f2 = 3 # inlined structure field access
+        ex.mystruct = MyStructClass(vm, addr2) # struct copy
+
+    It can be seen like a bridge to use a MemStruct as a MemField
+
+    TODO: make the Inline implicit when setting a field to be a MemStruct
+    """
+
+    def __init__(self, inlined_type, *type_args, **type_kwargs):
+        if not issubclass(inlined_type, MemStruct):
+            raise ValueError("inlined type if Inline must be a MemStruct")
+        self._il_type = inlined_type
+        self._type_args = type_args
+        self._type_kwargs = type_kwargs
+
+    def set(self, vm, addr, val):
+        raw = str(val)
+        vm.set_mem(addr, raw)
+
+    def get(self, vm, addr):
+        return self._il_type(vm, addr)
+
+    def size(self):
+        return self._il_type.sizeof()
+
+    def __repr__(self):
+        return "%s(%r)" % (self.__class__.__name__, self._il_type)
+
+    def __eq__(self, other):
+        return self.__class__ == other.__class__ and \
+                self._il_type == other._il_type and \
+                self._type_args == other._type_args and \
+                self._type_kwargs == other._type_kwargs
+
+    def __hash__(self):
+        return hash(hash(self.__class__) + hash(self._il_type) +
+                hash(self._type_args))
+
+
+class Array(MemField):
+    """A fixed size array (contiguous sequence) of a MemField subclass
+    elements. Similar to something like the char[10] type in C.
+
+    Getting an array field actually returns a MemSizedArray. Setting it is
+    possible with either a list or a MemSizedArray instance. Examples of syntax:
+
+        class Example(MemStruct):
+            fields = [("array", Array(Num("B"), 4))]
+
+        mystruct = Example(vm, addr)
+        mystruct.array[3] = 27
+        mystruct.array = [1, 4, 8, 9]
+        mystruct.array = MemSizedArray(vm, addr2, Num("B"), 4)
+    """
+
+    def __init__(self, field_type, array_len):
+        self.field_type = field_type
+        self.array_len = array_len
+
+    def _set_self_type(self, self_type):
+        super(Array, self)._set_self_type(self_type)
+        self.field_type._set_self_type(self_type)
+
+    def set(self, vm, addr, val):
+        # MemSizedArray assignment
+        if isinstance(val, MemSizedArray):
+            if val.array_len != self.array_len or len(val) != self.size():
+                raise ValueError("Size mismatch in MemSizedArray assignment")
+            raw = str(val)
+            vm.set_mem(addr, raw)
+
+        # list assignment
+        elif isinstance(val, list):
+            if len(val) != self.array_len:
+                raise ValueError("Size mismatch in MemSizedArray assignment ")
+            offset = 0
+            for elt in val:
+                self.field_type.set(vm, addr + offset, elt)
+                offset += self.field_type.size()
+
+        else:
+            raise NotImplementedError(
```

Comment:
Even if it is not the case in the rest of Miasm code, `NotImplementedError` should be reserved for abstract method.
When a method is not implemented, you can raise a `RuntimeError`, for instance.

[Source](https://docs.python.org/2/library/exceptions.html):
```
 exception NotImplementedError

    This exception is derived from RuntimeError. In user defined base classes, abstract methods should raise this exception when they require derived classes to override the method.
```
We should probably create an exception in Miasm for this kind of case (`FdsException`? :smiley:)